rs/types/management_canister_types/src/lib.rs

```diff
@@ -808,6 +834,106 @@ impl From<pb_canister_state_bits::LogVisibility> for LogVisibility {
     }
 }
 
+/// Maximum number of allowed log viewers (specified in the interface spec).
+const MAX_ALLOWED_LOG_VIEWERS_COUNT: usize = 10;
+
+pub type BoundedAllowedViewers =
+    BoundedVec<MAX_ALLOWED_LOG_VIEWERS_COUNT, UNBOUNDED, UNBOUNDED, PrincipalId>;
+
+/// Log visibility for a canister.
+/// ```text
+/// variant {
+///    controllers;
+///    public;
+///    allowed_viewers: vec principal;
+/// }
+/// ```
+#[derive(Default, Clone, CandidType, Deserialize, Debug, PartialEq, Eq, EnumIter)]
+pub enum LogVisibilityV2 {
+    #[default]
+    #[serde(rename = "controllers")]
+    Controllers,
+    #[serde(rename = "public")]
+    Public,
+    #[serde(rename = "allowed_viewers")]
+    AllowedViewers(BoundedAllowedViewers),
+}
+
+// Ensure backward compatibility with the old `LogVisibility`.
+// First read the new `LogVisibilityV2`, if it fails, try to read the old `LogVisibility`.
+impl Payload<'_> for LogVisibilityV2 {
+    fn decode(blob: &'_ [u8]) -> Result<Self, UserError> {
+        let args = match Decode!([decoder_config()]; blob, Self).map_err(candid_error_to_user_error)
+        {
+            Ok(record) => match record {
+                // TODO(EXC-1670): ignore receiving `AllowedViewers` variant, fallback to default.
+                Self::AllowedViewers(_) => Self::default(),
+                _ => record,
+            },
+            Err(_) => LogVisibility::decode(blob)?.into(),
```

Comment:
Is the fallback in this case really necessary? I would expect that `LogVisibilityV2` would always be able to decode a `LogVisibilityV1` as well.

Reply:
Agreed. Either `LogVisibilityV2` is bitwise backwards compatible with `LogVisibility`; or (assuming they're not) then trying to decode either of them as the other could theoretically result in accidentally interpreting some variant as some other variant.

I would rather make the explicit assumption (and ideally have a test for it) that `LogVisibilityV2` can always correctly decode a `LogVisibility`-encoded value.

Reply:
Ack, removed this conversion. Added corresponding deserialization tests for reading `LogVisibility` as `LogVisibilityV2`.